File: server/services/emailService.js
import nodemailer from 'nodemailer'
import EmailSettings from '../models/EmailSettings.js'
import EmailTemplate from '../models/EmailTemplate.js'

// Refresh OAuth2 access token
const refreshOAuth2Token = async (clientId, clientSecret, refreshToken) => {
  try {
    const response = await fetch('https://login.microsoftonline.com/common/oauth2/v2.0/token', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      body: new URLSearchParams({
        client_id: clientId,
        client_secret: clientSecret,
        refresh_token: refreshToken,
        grant_type: 'refresh_token',
        scope: 'https://outlook.office.com/SMTP.Send https://outlook.office.com/IMAP.AccessAsUser.All offline_access',
      }),
    })

    if (!response.ok) {
      throw new Error(`Token refresh failed: ${response.statusText}`)
    }

    const data = await response.json()
    return {
      accessToken: data.access_token,
      expiresAt: new Date(Date.now() + (data.expires_in * 1000)),
      refreshToken: data.refresh_token || refreshToken,
    }
  } catch (error) {
    console.error('OAuth2 token refresh error:', error)
    throw error
  }
}

// Get valid OAuth2 access token (refresh if needed)
const getValidAccessToken = async (settings, type = 'smtp') => {
  const auth = type === 'smtp' ? settings.smtp.auth : settings.imap.auth
  
  if (!auth?.oauth2?.enabled) {
    return null
  }

  const oauth2 = auth.oauth2
  const now = new Date()
  
  // Check if token is expired or will expire soon (within 5 minutes)
  if (!oauth2.accessToken || !oauth2.expiresAt || oauth2.expiresAt <= new Date(now.getTime() + 5 * 60 * 1000)) {
    console.log('Refreshing OAuth2 access token...')
    const tokenData = await refreshOAuth2Token(
      oauth2.clientId,
      oauth2.clientSecret,
      oauth2.refreshToken
    )
    
    // Update settings with new token
    oauth2.accessToken = tokenData.accessToken
    oauth2.expiresAt = tokenData.expiresAt
    oauth2.refreshToken = tokenData.refreshToken
    
    const emailSettings = await EmailSettings.getSettings()
    if (type === 'smtp') {
      emailSettings.smtp.auth.oauth2 = oauth2
    } else {
      emailSettings.imap.auth.oauth2 = oauth2
    }
    await emailSettings.save()
    
    return tokenData.accessToken
  }
  
  return oauth2.accessToken
}

// Create SMTP transporter
const createTransporter = async () => {
  const settings = await EmailSettings.getSettings()
  
  if (!settings.smtp.enabled) {
    throw new Error('SMTP is not enabled. Please configure email settings.')
  }

  // Check if Office365/Outlook
  const isOffice365 = settings.smtp.host && (
    settings.smtp.host.includes('office365.com') || 
    settings.smtp.host.includes('outlook.com') ||
    settings.smtp.host.includes('office.com')
  )

  const transporterConfig = {
    host: settings.smtp.host,
    port: parseInt(settings.smtp.port),
    secure: false, // Office365 uses STARTTLS, not direct SSL
    pool: false,
    maxConnections: 1,
    maxMessages: 1,
  }

  // Configure authentication (OAuth2 or password)
  if (settings.smtp.auth?.oauth2?.enabled) {
    // OAuth2 authentication
    const accessToken = await getValidAccessToken(settings, 'smtp')
    const trimmedUser = settings.smtp.auth.user ? settings.smtp.auth.user.trim() : ''
    
    transporterConfig.auth = {
      type: 'OAuth2',
      user: trimmedUser,
      clientId: settings.smtp.auth.oauth2.clientId,
      clientSecret: settings.smtp.auth.oauth2.clientSecret,
      refreshToken: settings.smtp.auth.oauth2.refreshToken,
      accessToken: accessToken,
    }
  } else {
    // Password authentication
    const trimmedPassword = settings.smtp.auth.pass ? settings.smtp.auth.pass.trim() : ''
    const trimmedUser = settings.smtp.auth.user ? settings.smtp.auth.user.trim() : ''
    
    transporterConfig.auth = {
      user: trimmedUser,
      pass: trimmedPassword,
    }
  }

  // Add Office365-specific settings
  if (isOffice365) {
    transporterConfig.requireTLS = true
    transporterConfig.tls = {
      ciphers: 'SSLv3',
      rejectUnauthorized: false,
      minVersion: 'TLSv1.2',
    }
    // Office365 specific: use LOGIN method explicitly (unless OAuth2)
    if (!settings.smtp.auth?.oauth2?.enabled) {
      transporterConfig.authMethod = 'LOGIN'
    }
    transporterConfig.service = 'smtp.office365.com'
    transporterConfig.connectionTimeout = 60000
    transporterConfig.greetingTimeout = 30000
    transporterConfig.socketTimeout = 60000
  } else {
    // For other providers, use secure if port is 465
    transporterConfig.secure = settings.smtp.secure === true || settings.smtp.port === 465
  }

  return nodemailer.createTransport(transporterConfig)
}

// Send email
export const sendEmail = async (to, subject, html, text = null) => {
  try {
    const settings = await EmailSettings.getSettings()
    const transporter = await createTransporter()
    
    const fromEmail = settings.smtp.fromEmail || settings.smtp.auth.user
    const fromName = settings.smtp.fromName || 'Ticketing Tool'

    const mailOptions = {
      from: `"${fromName}" <${fromEmail}>`,
      to,
      subject,
      html,
      text: text || html.replace(/<[^>]*>/g, ''), // Strip HTML for text version
    }

    const info = await transporter.sendMail(mailOptions)
    console.log('Email sent:', info.messageId)
    return { success: true, messageId: info.messageId }
  } catch (error) {
    console.error('Error sending email:', error)
    throw error
  }
}

// Render template with variables
const renderEmailTemplate = (template, data) => {
  let rendered = template
  Object.keys(data).forEach(key => {
    const value = data[key]
    const regex = new RegExp(`\\{\\{${key}\\}\\}`, 'g')
    rendered = rendered.replace(regex, value !== null && value !== undefined ? String(value) : '')
  })
  return rendered
}

// Send ticket acknowledgment email
export const sendTicketAcknowledgment = async (ticket, customerEmail) => {
  // Check for custom template
  let customTemplate = null
  try {
    customTemplate = await EmailTemplate.findOne({
      type: 'ticket-created',
      isActive: true,
      $or: [
        { organization: ticket.organization || null },
        { organization: null }, // Global template
      ],
    }).sort({ organization: -1 }) // Prefer org-specific over global
  } catch (error) {
    console.error('Error fetching custom template:', error)
  }

  let subject = `Ticket #${ticket.ticketId} Created - ${ticket.title}`
  let html = ''
  
  // Get priority badge color
  const getPriorityColor = (priority) => {
    const colors = {
      'low': '#6b7280',
      'medium': '#3b82f6',
      'high': '#f59e0b',
      'urgent': '#ef4444',
    }
    return colors[priority] || '#3b82f6'
  }

  const priorityColor = getPriorityColor(ticket.priority)
  
  html = `
    <!DOCTYPE html>
    <html>
    <head>
      <meta charset="UTF-8">
      <meta name="viewport" content="width=device-width, initial-scale=1.0">
      <style>
        * { margin: 0; padding: 0; box-sizing: border-box; }
        body { 
          font-family: 'Segoe UI', Tahoma, Geneva, Verdana, sans-serif; 
          line-height: 1.6; 
          color: #333; 
          background-color: #f5f5f5;
          padding: 20px;
        }
        .email-container { 
          max-width: 600px; 
          margin: 0 auto; 
          background-color: #ffffff;
          border-radius: 8px;
          overflow: hidden;
          box-shadow: 0 2px 4px rgba(0,0,0,0.1);
        }
        .header-banner { 
          background: linear-gradient(135deg, #3b82f6 0%, #2563eb 100%); 
          color: white; 
          padding: 30px 20px; 
          text-align: center; 
        }
        .header-banner h1 {
          font-size: 28px;
          font-weight: 600;
          margin: 0;
          letter-spacing: 0.5px;
        }
        .content { 
          padding: 30px 20px; 
          background: #ffffff;
        }
        .greeting {
          font-size: 16px;
          color: #1f2937;
          margin-bottom: 15px;
        }
        .intro-text {
          font-size: 15px;
          color: #4b5563;
          margin-bottom: 25px;
          line-height: 1.7;
        }
        .ticket-details { 
          background: #f9fafb; 
          padding: 20px; 
          margin: 25px 0; 
          border-radius: 6px;
          border: 1px solid #e5e7eb;
        }
        .ticket-details h3 {
          font-size: 18px;
          color: #1f2937;
          margin-bottom: 15px;
          font-weight: 600;
        }
        .detail-row {
          margin-bottom: 12px;
          font-size: 14px;
        }
        .detail-row:last-child {
          margin-bottom: 0;
        }
        .detail-label {
          font-weight: 600;
          color: #374151;
          display: inline-block;
          min-width: 120px;
        }
        .detail-value {
          color: #1f2937;
        }
        .priority-badge {
          display: inline-block;
          padding: 4px 12px;
          border-radius: 12px;
          font-size: 12px;
          font-weight: 600;
          text-transform: uppercase;
          background-color: ${priorityColor}20;
          color: ${priorityColor};
        }
        .status-badge {
          display: inline-block;
          padding: 4px 12px;
          border-radius: 12px;
          font-size: 12px;
          font-weight: 600;
          text-transform: uppercase;
          background-color: #10b98120;
          color: #10b981;
        }
        .description-box {
          background: #ffffff;
          padding: 15px;
          border-radius: 4px;
          border-left: 3px solid #3b82f6;
          margin-top: 10px;
          font-size: 14px;
          color: #4b5563;
          line-height: 1.6;
        }
        .footer-text {
          font-size: 14px;
          color: #6b7280;
          margin-top: 25px;
          line-height: 1.6;
        }
        .signature {
          margin-top: 20px;
          font-size: 14px;
          color: #1f2937;
        }
        .signature strong {
          color: #111827;
        }
        .footer-note {
          text-align: center;
          padding: 20px;
          background: #f9fafb;
          border-top: 1px solid #e5e7eb;
          font-size: 12px;
          color: #9ca3af;
        }
        @media only screen and (max-width: 600px) {
          .email-container {
            width: 100% !important;
          }
          .content {
            padding: 20px 15px !important;
          }
          .header-banner {
            padding: 25px 15px !important;
          }
          .header-banner h1 {
            font-size: 24px !important;
          }
        }
      </style>
    </head>
    <body>
      <div class="email-container">
        <div class="header-banner">
          <h1>Ticket Created Successfully</h1>
        </div>
        <div class="content">
          <p class="greeting">Dear Customer,</p>
          <p class="intro-text">Thank you for contacting us. We have received your request and created a support ticket for you.</p>
          
          <div class="ticket-details">
            <h3>Ticket Details:</h3>
            <div class="detail-row">
              <span class="detail-label">Ticket ID:</span>
              <span class="detail-value">#${ticket.ticketId}</span>
            </div>
            <div class="detail-row">
              <span class="detail-label">Title:</span>
              <span class="detail-value">${ticket.title}</span>
            </div>
            <div class="detail-row">
              <span class="detail-label">Category:</span>
              <span class="detail-value">${ticket.category}</span>
            </div>
            <div class="detail-row">
              <span class="detail-label">Priority:</span>
              <span class="detail-value"><span class="priority-badge">${ticket.priority}</span></span>
            </div>
            <div class="detail-row">
              <span class="detail-label">Status:</span>
              <span class="detail-value"><span class="status-badge">${ticket.status}</span></span>
            </div>
            <div class="detail-row">
              <span class="detail-label">Description:</span>
            </div>
            <div class="description-box">
              ${ticket.description.replace(/\n/g, '<br>')}
            </div>
          </div>
          
          <p class="footer-text">Our support team will review your ticket and respond as soon as possible. You will receive email notifications when there are updates to your ticket.</p>
          
          <p class="footer-text">If you have any additional information, please reply to this email and it will be added to your ticket.</p>
          
          <div class="signature">
            <strong>Best regards,</strong><br>
            Support Team
          </div>
        </div>
        <div class="footer-note">
          <p>This is an automated message. Please do not reply directly to this email.</p>
        </div>
      </div>
    </body>
    </html>
  `

  return await sendEmail(customerEmail, subject, html)
}

// Send ticket update email
export const sendTicketUpdateEmail = async (ticket, customerEmail, changes) => {
  // Check for custom template
  let customTemplate = null
  try {
    customTemplate = await EmailTemplate.findOne({
      type: 'ticket-updated',
      isActive: true,
      $or: [
        { organization: ticket.organization || null },
        { organization: null }, // Global template
      ],
    }).sort({ organization: -1 }) // Prefer org-specific over global
  } catch (error) {
    console.error('Error fetching custom template:', error)
  }

  let subject = `Ticket #${ticket.ticketId} Updated - ${ticket.title}`
  let html = ''
  
  // If custom template exists, use it
  if (customTemplate) {
    const templateData = {
      ticketId: ticket.ticketId,
      ticketTitle: ticket.title,
      ticketCategory: ticket.category,
      ticketPriority: ticket.priority,
      ticketStatus: ticket.status,
      customerName: ticket.creator?.name || 'Customer',
      assigneeName: ticket.assignee?.name || 'Unassigned',
      changes: changes ? Object.entries(changes).map(([k, v]) => `${k}: ${v}`).join(', ') : '',
    }
    subject = renderEmailTemplate(customTemplate.subject, templateData)
    html = renderEmailTemplate(customTemplate.htmlBody, templateData)
  } else {
    // Use default template
    const changesHtml = Object.entries(changes || {})
      .map(([key, value]) => `<li style="margin-bottom: 8px;"><strong>${key}:</strong> ${value}</li>`)
      .join('')
    
    // Get status color
    const getStatusColor = (status) => {
    const colors = {
      'open': '#3b82f6',
      'approval-pending': '#f59e0b',
      'approved': '#10b981',
      'rejected': '#ef4444',
      'in-progress': '#8b5cf6',
      'resolved': '#10b981',
      'closed': '#6b7280',
    }
    return colors[status] || '#3b82f6'
  }

  // Get priority color
  const getPriorityColor = (priority) => {
    const colors = {
      'low': '#6b7280',
      'medium': '#3b82f6',
      'high': '#f59e0b',
      'urgent': '#ef4444',
    }
    return colors[priority] || '#3b82f6'
  }

  const statusColor = getStatusColor(ticket.status)
  const priorityColor = getPriorityColor(ticket.priority)
  
  html = `
    <!DOCTYPE html>
    <html>
    <head>
      <meta charset="UTF-8">
      <meta name="viewport" content="width=device-width, initial-scale=1.0">
      <style>
        * { margin: 0; padding: 0; box-sizing: border-box; }
        body { 
          font-family: 'Segoe UI', Tahoma, Geneva, Verdana, sans-serif; 
          line-height: 1.6; 
          color: #333; 
          background-color: #f5f5f5;
          padding: 20px;
        }
        .email-container { 
          max-width: 600px; 
          margin: 0 auto; 
          background-color: #ffffff;
          border-radius: 8px;
          overflow: hidden;
          box-shadow: 0 2px 4px rgba(0,0,0,0.1);
        }
        .header-banner { 
          background: linear-gradient(135deg, ${statusColor} 0%, ${statusColor}dd 100%); 
          color: white; 
          padding: 30px 20px; 
          text-align: center; 
        }
        .header-banner h1 {
          font-size: 28px;
          font-weight: 600;
          margin: 0;
          letter-spacing: 0.5px;
        }
        .content { 
          padding: 30px 20px; 
          background: #ffffff;
        }
        .greeting {
          font-size: 16px;
          color: #1f2937;
          margin-bottom: 15px;
        }
        .intro-text {
          font-size: 15px;
          color: #4b5563;
          margin-bottom: 25px;
          line-height: 1.7;
        }
        .ticket-details { 
          background: #f9fafb; 
          padding: 20px; 
          margin: 25px 0; 
          border-radius: 6px;
          border: 1px solid #e5e7eb;
        }
        .ticket-details h3 {
          font-size: 18px;
          color: #1f2937;
          margin-bottom: 15px;
          font-weight: 600;
        }
        .detail-row {
          margin-bottom: 12px;
          font-size: 14px;
        }
        .detail-row:last-child {
          margin-bottom: 0;
        }
        .detail-label {
          font-weight: 600;
          color: #374151;
          display: inline-block;
          min-width: 120px;
        }
        .detail-value {
          color: #1f2937;
        }
        .priority-badge {
          display: inline-block;
          padding: 4px 12px;
          border-radius: 12px;
          font-size: 12px;
          font-weight: 600;
          text-transform: uppercase;
          background-color: ${priorityColor}20;
          color: ${priorityColor};
        }
        .status-badge {
          display: inline-block;
          padding: 4px 12px;
          border-radius: 12px;
          font-size: 12px;
          font-weight: 600;
          text-transform: uppercase;
          background-color: ${statusColor}20;
          color: ${statusColor};
        }
        .changes-box {
          background: #f0fdf4;
          padding: 20px;
          border-radius: 6px;
          border-left: 3px solid #10b981;
          margin: 25px 0;
        }
        .changes-box h3 {
          font-size: 18px;
          color: #1f2937;
          margin-bottom: 15px;
          font-weight: 600;
        }
        .changes-box ul {
          list-style: none;
          padding: 0;
          margin: 0;
        }
        .changes-box li {
          font-size: 14px;
          color: #4b5563;
          margin-bottom: 8px;
        }
        .view-button {
          display: inline-block;
          padding: 12px 24px;
          background: #3b82f6;
          color: white;
          text-decoration: none;
          border-radius: 6px;
          margin: 20px 0;
          font-weight: 600;
          font-size: 14px;
        }
        .view-button:hover {
          background: #2563eb;
        }
        .footer-text {
          font-size: 14px;
          color: #6b7280;
          margin-top: 25px;
          line-height: 1.6;
        }
        .signature {
          margin-top: 20px;
          font-size: 14px;
          color: #1f2937;
        }
        .signature strong {
          color: #111827;
        }
        .footer-note {
          text-align: center;
          padding: 20px;
          background: #f9fafb;
          border-top: 1px solid #e5e7eb;
          font-size: 12px;
          color: #9ca3af;
        }
        @media only screen and (max-width: 600px) {
          .email-container {
            width: 100% !important;
          }
          .content {
            padding: 20px 15px !important;
          }
          .header-banner {
            padding: 25px 15px !important;
          }
          .header-banner h1 {
            font-size: 24px !important;
          }
        }
      </style>
    </head>
    <body>
      <div class="email-container">
        <div class="header-banner">
          <h1>Ticket Updated</h1>
        </div>
        <div class="content">
          <p class="greeting">Dear Customer,</p>
          <p class="intro-text">Your ticket #${ticket.ticketId} has been updated.</p>
          
          <div class="ticket-details">
            <h3>Current Ticket Status:</h3>
            <div class="detail-row">
              <span class="detail-label">Ticket ID:</span>
              <span class="detail-value">#${ticket.ticketId}</span>
            </div>
            <div class="detail-row">
              <span class="detail-label">Title:</span>
              <span class="detail-value">${ticket.title}</span>
            </div>
            <div class="detail-row">
              <span class="detail-label">Status:</span>
              <span class="detail-value"><span class="status-badge">${ticket.status.toUpperCase().replace('-', ' ')}</span></span>
            </div>
            <div class="detail-row">
              <span class="detail-label">Priority:</span>
              <span class="detail-value"><span class="priority-badge">${ticket.priority}</span></span>
            </div>
            <div class="detail-row">
              <span class="detail-label">Assignee:</span>
              <span class="detail-value">${ticket.assignee?.name || 'Unassigned'}</span>
            </div>
            ${ticket.dueDate ? `
            <div class="detail-row">
              <span class="detail-label">Due Date:</span>
              <span class="detail-value">${new Date(ticket.dueDate).toLocaleString()}</span>
            </div>
            ` : ''}
          </div>
          
          ${changesHtml ? `
          <div class="changes-box">
            <h3>Recent Changes:</h3>
            <ul>${changesHtml}</ul>
          </div>
          ` : ''}
          
          <p class="footer-text">You will receive email notifications for all status updates on this ticket.</p>
          
          <p><a href="${process.env.FRONTEND_URL || 'http://localhost'}/tickets/${ticket.ticketId}" class="view-button">View Ticket</a></p>
          
          <div class="signature">
            <strong>Best regards,</strong><br>
            Support Team
          </div>
        </div>
        <div class="footer-note">
          <p>This is an automated message.</p>
        </div>
      </div>
    </body>
    </html>
    `
  }

  return await sendEmail(customerEmail, subject, html)
}

File: backend/app/services/email_service.py
"""
Email service for sending emails using configured SMTP settings
"""
import smtplib
from email.mime.text import MIMEText
from email.mime.multipart import MIMEMultipart
from typing import Optional
from app.db.database import get_database
from app.services.email_oauth2 import EmailOAuth2Service


async def send_email(
    to: str,
    subject: str,
    html: Optional[str] = None,
    text: Optional[str] = None
) -> bool:
    """
    Send email using configured SMTP settings
    
    Args:
        to: Recipient email address
        subject: Email subject
        html: HTML email body (optional)
        text: Plain text email body (optional, required if html not provided)
    
    Returns:
        bool: True if email sent successfully, False otherwise
    """
    db = await get_database()
    
    # Get email settings
    email_settings = await db.email_settings.find_one({})
    if not email_settings:
        print("WARNING: Email settings not configured. Cannot send email.")
        return False
    
    smtp_config = email_settings.get("smtp", {})
    if not smtp_config or not smtp_config.get("enabled", False):
        print("WARNING: SMTP is not enabled. Cannot send email.")
        return False
    
    auth_method = smtp_config.get("authMethod", "password")
    host = smtp_config.get("host")
    port = smtp_config.get("port", 587)
    encryption = smtp_config.get("encryption", "TLS")
    from_email = smtp_config.get("fromEmail") or smtp_config.get("auth", {}).get("user")
    from_name = smtp_config.get("fromName", "Ticketing Tool")
    
    if not host or not from_email:
        print("WARNING: SMTP host or from email not configured.")
        return False
    
    try:
        # Create message
        if html:
            msg = MIMEMultipart("alternative")
            msg.attach(MIMEText(text or "", "plain"))
            msg.attach(MIMEText(html, "html"))
        else:
            msg = MIMEText(text or "", "plain")
        
        msg["From"] = f"{from_name} <{from_email}>"
        msg["To"] = to
        msg["Subject"] = subject
        
        # Connect to SMTP server
        if auth_method == "oauth2":
            # OAuth2 authentication
            oauth2_config = smtp_config.get("auth", {}).get("oauth2", {})
            provider = oauth2_config.get("provider", "microsoft")
            
            # Get/refresh access token
            if provider == "microsoft":
                token_data = await EmailOAuth2Service.get_microsoft_token(
                    client_id=oauth2_config.get("clientId"),
                    client_secret=oauth2_config.get("clientSecret"),
                    tenant_id=oauth2_config.get("tenantId"),
                    refresh_token=oauth2_config.get("refreshToken")
                )
            else:  # google
                token_data = await EmailOAuth2Service.get_google_token(
                    client_id=oauth2_config.get("clientId"),
                    client_secret=oauth2_config.get("clientSecret"),
                    refresh_token=oauth2_config.get("refreshToken")
                )
            
            # Update token in settings
            smtp_config["auth"]["oauth2"]["accessToken"] = token_data["access_token"]
            smtp_config["auth"]["oauth2"]["refreshToken"] = token_data.get("refresh_token") or oauth2_config.get("refreshToken")
            smtp_config["auth"]["oauth2"]["tokenExpiry"] = token_data["token_expiry"]
            
            await db.email_settings.update_one(
                {},
                {"$set": {"smtp": smtp_config}}
            )
            
            # For OAuth2, we would need to use XOAUTH2 authentication
            # This is more complex and requires additional libraries
            # For now, fall back to password auth if OAuth2 tokens are not available
            print("WARNING: OAuth2 email sending not fully implemented. Using password auth if available.")
            auth_method = "password"
        
        if auth_method == "password":
            # Password authentication
            username = smtp_config.get("auth", {}).get("user")
            password = smtp_config.get("auth", {}).get("pass")
            
            if not username or not password:
                print("WARNING: SMTP username or password not configured.")
                return False
            
            # Connect and send
            if encryption == "SSL":
                server = smtplib.SMTP_SSL(host, port)
            else:
                server = smtplib.SMTP(host, port)
                if encryption == "TLS":
                    server.starttls()
            
            server.login(username, password)
            server.send_message(msg)
            server.quit()
            
            print(f"INFO: Email sent successfully to {to}")
            return True
            
    except Exception as e:
        print(f"ERROR: Failed to send email to {to}: {str(e)}")
        return False


async def send_welcome_email(
    user_email: str,
    user_name: str,
    password: Optional[str] = None,
    login_url: Optional[str] = None
) -> bool:
    """
    Send welcome email to newly created user
    
    Args:
        user_email: User's email address
        user_name: User's name
        password: Temporary password (if provided)
        login_url: Login page URL
    
    Returns:
        bool: True if email sent successfully
    """
    if not login_url:
        login_url = "http://localhost/login"
    
    subject = "Welcome to Ticketing Tool - Your Account Has Been Created"
    
    # HTML email template
    html_body = f"""
    <!DOCTYPE html>
    <html>
    <head>
        <meta charset="UTF-8">
        <style>
            body {{
                font-family: Arial, sans-serif;
                line-height: 1.6;
                color: #333;
                max-width: 600px;
                margin: 0 auto;
                padding: 20px;
            }}
            .header {{
                background-color: #4F46E5;
                color: white;
                padding: 20px;
                text-align: center;
                border-radius: 5px 5px 0 0;
            }}
            .content {{
                background-color: #f9fafb;
                padding: 30px;
                border: 1px solid #e5e7eb;
            }}
            .button {{
                display: inline-block;
                background-color: #4F46E5;
                color: white;
                padding: 12px 24px;
                text-decoration: none;
                border-radius: 5px;
                margin: 20px 0;
            }}
            .credentials {{
                background-color: #fff;
                border: 1px solid #e5e7eb;
                border-radius: 5px;
                padding: 15px;
                margin: 20px 0;
            }}
            .footer {{
                text-align: center;
                color: #6b7280;
                font-size: 12px;
                margin-top: 20px;
            }}
        </style>
    </head>
    <body>
        <div class="header">
            <h1>Welcome to Ticketing Tool!</h1>
        </div>
        <div class="content">
            <p>Hello {user_name},</p>
            
            <p>Your account has been successfully created in the Ticketing Tool system.</p>
            
            {f'''
            <div class="credentials">
                <p><strong>Your Login Credentials:</strong></p>
                <p><strong>Email:</strong> {user_email}</p>
                {f'<p><strong>Temporary Password:</strong> {password}</p>' if password else '<p><strong>Password:</strong> Use the password you set during registration</p>'}
            </div>
            
            <p><strong>Important:</strong> Please change your password after your first login for security purposes.</p>
            ''' if password else ''}
            
            <p>You can now access the Ticketing Tool by clicking the button below:</p>
            
            <div style="text-align: center;">
                <a href="{login_url}" class="button">Login to Ticketing Tool</a>
            </div>
            
            <p>If you have any questions or need assistance, please contact your system administrator.</p>
            
            <p>Best regards,<br>The Ticketing Tool Team</p>
        </div>
        <div class="footer">
            <p>This is an automated email. Please do not reply to this message.</p>
        </div>
    </body>
    </html>
    """
    
    # Plain text version
    text_body = f"""
Welcome to Ticketing Tool!

Hello {user_name},

Your account has been successfully created in the Ticketing Tool system.

{f'''
Your Login Credentials:
Email: {user_email}
Temporary Password: {password}

Important: Please change your password after your first login for security purposes.
''' if password else ''}

You can now access the Ticketing Tool at: {login_url}

If you have any questions or need assistance, please contact your system administrator.

Best regards,
The Ticketing Tool Team

---
This is an automated email. Please do not reply to this message.
    """
    
    return await send_email(
        to=user_email,
        subject=subject,
        html=html_body,
        text=text_body
    )
